src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
                 }
             }
         }
 
-        public void QueueDiagnostics(params string[] documents)
+        private IEnumerable<(Project project, CancellationTokenSource workReadySource)> GetThrottledWork()
         {
-            this.EmitDiagnostics(documents);
-        }
-
-        private void EmitDiagnostics(params string[] documents)
-        {
-            if (_forwarder.IsEnabled)
+            lock (_workQueue)
             {
-                foreach (var document in documents)
-                {
-                    if (!_openDocuments.Contains(document))
-                    {
-                        _openDocuments.Enqueue(document);
-                    }
-                }
+                var currentWork = _workQueue
+                    .Where(x => x.Value.modified.AddMilliseconds(_throttlingMs) < DateTime.UtcNow)
+                    .OrderByDescending(x => x.Value.modified) // If you currently edit project X you want it will be highest priority and contains always latest possible analysis.
+                    .Take(2) // Limit mount of work executed by once. This is needed on large solution...
+                    .ToList();
 
-                if (!_queueRunning && !_openDocuments.IsEmpty)
+                foreach (var workKey in currentWork.Select(x => x.Key))
                 {
-                    this.ProcessQueue();
+                    _workQueue.TryRemove(workKey, out _);
                 }
+
+                return currentWork.Select(x => (x.Value.project, x.Value.workReadySource));
             }
         }
 
-        private void ProcessQueue()
+        public async Task<IEnumerable<(string projectName, Diagnostic diagnostic)>> GetCurrentDiagnosticResult(IEnumerable<ProjectId> projectIds)
+        {
+            await WaitForInitialStartupWorkIfAny();
+
+            var pendingWork = WaitForPendingWorkIfNeededAndGetIt(projectIds);
+
+            await Task.WhenAll(pendingWork);
+
+            return _results
+                .Where(x => projectIds.Any(pid => pid == x.Key))
+                .SelectMany(x => x.Value.diagnostics, (k, v) => ((k.Value.name, v)));
+        }
+
+        public void QueueForAnalysis(IEnumerable<Project> projects)
         {
-            lock (_lock)
+            foreach (var project in projects)
             {
-                _queueRunning = true;
+                _workQueue.AddOrUpdate(project.Id,
+                    (modified: DateTime.UtcNow, project: project, new CancellationTokenSource()),
+                    (_, oldValue) => (modified: DateTime.UtcNow, project: project, oldValue.workReadySource));
             }
+        }
 
-            Task.Factory.StartNew(async () =>
-            {
-                await Task.Delay(100);
-                await Dequeue();
+        private IEnumerable<Task> WaitForPendingWorkIfNeededAndGetIt(IEnumerable<ProjectId> projectIds)
+        {
+            return _workQueue
```

Comment:
curious what it is trying to do?

Reply:
Most of complexity of diagnostic service comes from how omnisharp api (soon legacy) is used, consumers seems to kind of poll results several times and since analyzers are eventually ready these are workaround for that problem. LSP with event based diagnostic will fix this issue and i think this complexity can be removed then. Original version just executed analysis and cached them, rest of logig came afterwads when i battled with user experience with editors how they surface diagnostic via polling api. However theres possibility that i just got it completly wrong, but this is on my current understanding how api is used.

Reply:
regardless how ominisharp does, I believe you should make sure that document (the solution snapshot got used) used to create diagnostics and fix must be same one. otherwise, you will get random exception from fixer.

ex) fixer gets unknown type diagnostics from analyzer, it got fed into generate type fixer with a document that no longer has that diagnostic, fixer does find token on the location of the diagnostic since it assumes the diagnostics is from the same document it got so it must exist. but it doesn't. so find token return random token which the fixer tries to cast to identifier, and fix throw invalid cast exception.

it is just 1 example. all kind of random thing will happen if you feed in some random diagnostics from some random old version of text from same file.

Reply:
I wouldn't call omnisharp API "soon legacy". The LSP implementation doesn't provide all the features used in the C# extension and we don't currently have plans to move away from it (and I assume that's the case for other O# consumers too)